### PR TITLE
[SPARK-17304] Fix perf. issue caused by TaskSetManager.abortIfCompletelyBlacklisted

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch addresses a minor scheduler performance issue that was introduced in #13603. If you run

```
sc.parallelize(1 to 100000, 100000).map(identity).count()
```

then most of the time ends up being spent in `TaskSetManager.abortIfCompletelyBlacklisted()`:

image

When processing resource offers, the scheduler uses a nested loop which considers every task set at multiple locality levels:

``` scala
   for (taskSet <- sortedTaskSets; maxLocality <- taskSet.myLocalityLevels) {
      do {
        launchedTask = resourceOfferSingleTaskSet(
            taskSet, maxLocality, shuffledOffers, availableCpus, tasks)
      } while (launchedTask)
    }
```

In order to prevent jobs with globally blacklisted tasks from hanging, #13603 added a `taskSet.abortIfCompletelyBlacklisted` call inside of  `resourceOfferSingleTaskSet`; if a call to `resourceOfferSingleTaskSet` fails to schedule any tasks, then `abortIfCompletelyBlacklisted` checks whether the tasks are completely blacklisted in order to figure out whether they will ever be schedulable. The problem with this placement of the call is that the last call to `resourceOfferSingleTaskSet` in the `while` loop will return `false`, implying that  `resourceOfferSingleTaskSet` will call `abortIfCompletelyBlacklisted`, so almost every call to `resourceOffers` will trigger the `abortIfCompletelyBlacklisted` check for every task set.

Instead, I think that this call should be moved out of the innermost loop and should be called _at most_ once per task set in case none of the task set's tasks can be scheduled at any locality level.

Before this patch's changes, the microbenchmark example that I posted above took 35 seconds to run, but it now only takes 15 seconds after this change.

/cc @squito and @kayousterhout for review.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 1]
```diff
@@ -326,12 +323,19 @@ private[spark] class TaskSchedulerImpl(
     // Take each TaskSet in our scheduling order, and then offer it each node in increasing order
     // of locality levels so that it gets a chance to launch local tasks on all of them.
     // NOTE: the preferredLocality order: PROCESS_LOCAL, NODE_LOCAL, NO_PREF, RACK_LOCAL, ANY
-    var launchedTask = false
-    for (taskSet <- sortedTaskSets; maxLocality <- taskSet.myLocalityLevels) {
-      do {
-        launchedTask = resourceOfferSingleTaskSet(
+    for (taskSet <- sortedTaskSets) {
+      var launchedAnyTask = false
+      var launchedTaskAtMaxLocality = false
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: I know it's super verbose, but to minimize how confusing this code is, what about calling this launchedTaskAtCurrentMaxLocality, and renaming maxLocality below to currentMaxLocality?

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Good idea; updated.

**Comment by kayousterhout** [comment]
LGTM

Josh, how long does your microbenchmark take if you comment out the call to abortIfCompletelyBlacklisted? Wondering how much that continues to affect performance.

**Comment by JoshRosen** [comment]
@kayousterhout, after this patch's changes commenting out `abortIfCompletelyBlacklisted` appears to provide no further performance improvement. However, I have identified a few bottlenecks in Accumulator serialization that can be used to obtain an additional ~30% speedup on this microbenchmark; I'll submit those in a separate patch.

**Comment by kayousterhout** [comment]
Ok cool thanks @JoshRosen -- just wanted to make sure that the blacklisting wasn't still hurting performance.  Thanks for fixing this and sorry about the oversight originally!

**Comment by JoshRosen** [comment]
Merging this to master. Thanks!

**Comment by squito** [comment]
thanks for finding this and the quick fix @JoshRosen !

**Pull request closed without merging**